src/transformers/models/maskformer/modeling_maskformer.py

```diff
@@ -0,0 +1,1638 @@
+
+
+    Returns:
+        Tensor: The computed loss between each pairs
+    """
+    inputs: Tensor = inputs.sigmoid()
+    # TODO this .flatten seems to be unecessary because the shape is 2d
+    inputs: Tensor = inputs.flatten(1)
+    # TODO why 1 is not added to the number to avoid numerator = 0 in edge cases?
+    numerator: Tensor = 2 * torch.einsum("nc,mc->nm", inputs, targets)
+    # using broadcasting to get a [NUM_QUERIES, NUM_CLASSES] matrix
+    denominator: Tensor = inputs.sum(-1)[:, None] + targets.sum(-1)[None, :]
+    loss: Tensor = 1 - (numerator + 1) / (denominator + 1)
+    return loss
+
+
+def pair_wise_sigmoid_focal_loss(inputs: Tensor, targets: Tensor, alpha: float = 0.25, gamma: float = 2.0) -> Tensor:
+    """
+    A pair wise version of the focal loss, see `sigmoid_focal_loss` for usage
+
+    Args:
+        inputs (Tensor): A tensor representing a mask
+        targets (Tensor): A tensor with the same shape as inputs. Stores the binary classification labels for each element in inputs (0 for the negative class and 1 for the positive class).
+
+
+    Returns:
+        Tensor: The computed loss between each pairs
+    """
+    if alpha < 0:
+        raise ValueError(f"alpha must be positive")
+
+    hw: int = inputs.shape[1]
+
+    prob: Tensor = inputs.sigmoid()
+    focal_pos: Tensor = ((1 - prob) ** gamma) * F.binary_cross_entropy_with_logits(
+        inputs, torch.ones_like(inputs), reduction="none"
+    )
+    focal_neg: Tensor = (prob ** gamma) * F.binary_cross_entropy_with_logits(
+        inputs, torch.zeros_like(inputs), reduction="none"
+    )
+
+    focal_pos *= alpha
+    focal_neg *= 1 - alpha
+
+    loss: Tensor = torch.einsum("nc,mc->nm", focal_pos, targets) + torch.einsum("nc,mc->nm", focal_neg, (1 - targets))
+
+    return loss / hw
+
+
+# refactored from original implementation
+class MaskFormerHungarianMatcher(nn.Module):
+    """This class computes an assignment between the targets and the predictions of the network
+
+    For efficiency reasons, the targets don't include the no_object. Because of this, in general,
+    there are more predictions than targets. In this case, we do a 1-to-1 matching of the best predictions,
+    while the others are un-matched (and thus treated as non-objects).
+    """
+
+    def __init__(self, cost_class: float = 1.0, cost_mask: float = 1.0, cost_dice: float = 1.0):
+        """Creates the matcher
+
+        Params:
+            cost_class: This is the relative weight of the classification error in the matching cost
+            cost_mask: This is the relative weight of the focal loss of the binary mask in the matching cost
+            cost_dice: This is the relative weight of the dice loss of the binary mask in the matching cost
+        """
+        super().__init__()
+        assert cost_class != 0 or cost_mask != 0 or cost_dice != 0, "all costs cant be 0"
```

Comment:
No `assert` in a modeling file. We use tests and raise errors (as asserts can be skipped in some envs).